substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/Resources.java

```diff
@@ -99,19 +107,28 @@ public static byte[] inputStreamToByteArray(InputStream is) {
         }
     }
 
+    private static void updateTimeStamp() {
+        if (singleton().lastModifiedTime == INVALID_TIMESTAMP) {
+            singleton().lastModifiedTime = new Date().getTime();
+        }
+    }
+
+    private static Object addEntry(String moduleName, String resourceName, Object newEntry, EconomicMap<Pair<String, String>, Object> resources) {
+        Pair<String, String> key = Pair.create(moduleName, resourceName);
+        Object entry = resources.get(key);
+        if (entry == null || entry == NEGATIVE_QUERY) {
+            entry = newEntry;
+            updateTimeStamp();
+            resources.put(key, newEntry);
+        }
+        return entry;
+    }
+
     private static void addEntry(String moduleName, String resourceName, boolean isDirectory, byte[] data, boolean fromJar) {
         var resources = singleton().resources;
         synchronized (resources) {
-            Pair<String, String> key = Pair.create(moduleName, resourceName);
-            ResourceStorageEntry entry = resources.get(key);
-            if (entry == null) {
-                if (singleton().lastModifiedTime == INVALID_TIMESTAMP) {
-                    singleton().lastModifiedTime = new Date().getTime();
-                }
-                entry = new ResourceStorageEntry(isDirectory, fromJar);
-                resources.put(key, entry);
-            }
-            entry.getData().add(data);
+            Object entry = addEntry(moduleName, resourceName, new ResourceStorageEntry(isDirectory, fromJar), resources);
```

Comment:
This computes the new entry even if it is not used, just pass the constructor arguments and build the `ResourceStorageEntry` on-demand.